enhancements/network/multi-networkpolicy.md

```diff
@@ -0,0 +1,178 @@
+---
+title: multi-networkpolicy
+authors:
+  - "@s1061123"
+reviewers:
+  - @dcbw
+  - @dougbtv
+  - @fepan
+  - @zshi-redhat
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-08-07
+last-updated: yyyy-mm-dd
+status: implementable
+---
+
+# NetworkPolicy for Multus Interfaces
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal to bring networkpolicy features for multus interface.
+
+## Motivation
+
+Customer wants to make their Kubernetes secure, by filtering traffic from
+unexpected source. Kubernetes provides `NetworkPolicy` feature, however, it
+is only applied to default network interface, not additional network interface
+that is created by Multus CNI.  NetworkPolicy for Multus Interfaces brings the NetworkPolicy
+mechanism to additional network interfaces and it provides customers the ability to filter
+traffic from unexpected sources.
+
+### Goals
+
+- Introduce MultiNetworkPolicy CRD
+- Introduce multi-networkpolicy which generates iptables from CRD as shown in the Design Details section
+- Provide iptable based filtering mechanism for pod
+- Gathering related information (iptables rules in pod network namespace) by must-gather
+
+### Non-Goals
+
+- Provide eBPF filtering mechanism
+- Provide NetworkPolicy for userspace pod which uses DPDK and so on
+
+## Proposal
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Story 1
+
+Customer can specify how groups of pods are allowed to communicate with each other
+and other network endpoints as Kubernetes NetworkPolicy does.
```

Comment:
> It would be useful to specify in more detail what customer requests we've had, and any specific things we know that customers want.

> eg, one question not answered in this doc is: do we actually need different policies on different interfaces, or would it be sufficient to implement the same policies on every interface? If the answer to that is that we really need different policies on different interfaces then it would be useful to have a real-world example of a situation in which a customer wants "pod A can connect to pod B via net1 but not via eth0" or the like.

I believe that customer needs different policies on different interface because that's why customer want to have more than two network interface to the pod. For example, let's imagine that telecom customer uses multus as following for 5G mobile network:

- `eth0` connects to k8s cluster network
- `net1` connects to company management network (for telemetry)
- `net2` connects to 5G control plane network
- `net3` connects to 5G user plane network

In this case, of course, each might have different policy because each networks use-case is different. Telemetry just want to send the traffic to collector but 5G control plane have different network policy as well as 5G data plane does.

On the other side, customer may want to want to same policies on every interface, but it is not always, as above case.
As a platform vendor, I believe that we should be flexible to catch both use-cases, hence we need to provide to set different policies on different interfaces (i.e. networks).

> Another unanswered question is: do customers really want something NetworkPolicy-like, or would they rather have something more EgressFirewall / iptables-like? Do we actually expect them to use podSelectors and namespaceSelectors in the MultiNetworkPolicies, or is it all going to be CIDR-based?

>To the extent that it makes sense to talk about "cluster administrators" and "developers" as separate users of these clusters, is this feature targeted at the cluster administrators ("I want to protect my cluster from all users") or developers ("I want to protect my application from other users"). NetworkPolicy works by doing "deny everything by default, then allow only certain things", which works for the developer use case but tends to not work very well for the cluster-admin use case (which is why EgressFirewall works totally differently).

I suppose that it is depends on user. If user is good at iptable-like one, user want to have it. I believe that OpenShift user is good at Kubernetes because OpenShift is Kubernetes based. Hence I also believe that user may have some experiences of Kubernetes NetworkPolicy. That is the main reason to use NetworkPolicy mechanism to net-attach-def interfaces. Fortunately Kubernetes NetworkPolicy supports CIDR-based annotation as ipBlock and MultiNetworkPolicy also have ipBlock, so both can be supported.

If user want to have different mechanism to filter it, then we could implement it as different one because this MultiNetworkPolicy does not cover all use-cases of container network security.

> Hm... I see that below you said "I don't get real customer requirements yet"... If we don't have real customer requirements yet, then how do we know that this solution is going to work for them?

I propose this enhancement because customer want to have some security mechanism for net-attach-def network. If customer want NOT to have security, I suppose we should NOT provide this enhancement because it is useless. I believe that security is obviously required as the history of computer network security, hence I propose this enhancement even though we don't have specific customer requirement in detail (roughly we get request from some of telecom, but not detail use-cases). In future enhancement, we can make it better, i.e. more feasible to user case, by getting feedback from user as we update apiVersion with additional fields.

Enhancement template just said `Detail the things that people will be able to do if this is implemented.` in user story section, hence I just write as it is.
If we need to add detail use-cases with specific customer case, let's remove user story section because we don't have it yet and 'user story' is optional.

Reply:
> Enhancement template just said `Detail the things that people will be able to do if this is implemented.` in user story section, hence I just write as it is.
> If we need to add detail use-cases with specific customer case, let's remove user story section because we don't have it yet and 'user story' is optional.

I'm not worried about whether we are filling out the enhancement template correctly, I'm worried about whether we are actually implementing the feature our customers need or not.

NetworkPolicy was designed to work in a very specific way for a very specific use case. In general, I think the addition of `egress` and `ipBlock` to NetworkPolicy was a failure, and a lot of users who are trying to use those features are not especially happy with their limitations. (eg, there's no way to have one policy that says "block connections to 10.0.1.0/24" and one policy that says "block connections to 10.0.2.0/24" and then apply both of those policies to a single pod; if you try, you'll end up with an effective policy that says "allow all connections to everywhere").

If we think users will mostly be using MultiNetworkPolicy to create ingress rules based on podSelectors and namespaceSelectors, then that should work fine. But if we think they're mostly going to be doing stuff with egress or CIDR-based policies, then basing the API on NetworkPolicy is probably not the right plan.

Reply:
I got it. Currently I suppose that there is no policy language which satisfies customer requirements yet and this enhancement proposal does not challenge it, too. This proposal is a conservative approach to provide a network policy because this just provide same mechanism of Kubernetes (even tough we understand many user does not love it). Bad things is it is not perfect, however, good things is 'following upstream, Kubernetes' and it is very common. User does not have to know more format...

NetworkPolicy V2 is now discussed in the community, so I hope that new NetworkPolicy will improve it somehow. Then I am going to follow it.